src/ninja.cc

```diff
@@ -505,6 +507,213 @@ int NinjaMain::ToolDeps(const Options* options, int argc, char** argv) {
+    abs_gen_dir_ = buf;
+  }
+}
+
+bool MissingDependencyScanner::IsNodeInGenDir(Node* node) {
+  if (abs_gen_dir_.empty())
+    return false;
+  string path = node->path();
+  char abs_path_buf[PATH_MAX];
+  char* rp = realpath(path.c_str(), abs_path_buf);
+  return (rp && (strlen(rp) > abs_gen_dir_.size())) &&
+         (memcmp(rp, abs_gen_dir_.c_str(), abs_gen_dir_.size()) == 0);
+}
+
+void MissingDependencyScanner::ProcessNode(Node* node) {
+  if (!node)
+    return;
+  Edge* edge = node->in_edge();
+  if (!edge)
+    return;
+  if (!seen_.insert(node).second)
+    return;
+
+  for (vector<Node*>::iterator in = edge->inputs_.begin();
+       in != edge->inputs_.end(); ++in) {
+    ProcessNode(*in);
+  }
+
+  string deps_type = edge->GetBinding("deps");
+  if (!deps_type.empty()) {
+    DepsLog::Deps* deps = deps_log_->GetDeps(node);
+    if (deps)
+      ProcessNodeDeps(node, deps->nodes, deps->node_count);
+  } else {
+    ImplicitDepLoader dep_loader(state_, deps_log_, disk_interface_);
+    vector<Node*> depfile_deps;
+    string depfile = edge->GetUnescapedDepfile();
+    if (depfile.empty())
+      return;
+    string err;
+    dep_loader.LoadDepfileDeps(edge, depfile, &depfile_deps, &err);
+    if (!depfile_deps.empty())
+      ProcessNodeDeps(node, &depfile_deps[0], depfile_deps.size());
+  }
+}
+
+void MissingDependencyScanner::ProcessNodeDeps(Node* node, Node** dep_nodes,
+                                               int dep_nodes_count) {
+  Edge* edge = node->in_edge();
+  set<Edge*> deplog_edges;
+  vector<Node*> deplog_nodes_without_generator;
+  for (int i = 0; i < dep_nodes_count; ++i) {
+    Node* deplog_node = dep_nodes[i];
+    Edge* deplog_edge = deplog_node->in_edge();
+    if (deplog_edge) {
+      deplog_edges.insert(deplog_edge);
+    } else if (IsNodeInGenDir(deplog_node)) {
+      nodes_missing_gen_.insert(deplog_node);
+      deplog_nodes_without_generator.push_back(deplog_node);
+    }
+  }
+  vector<Edge*> missing_deps;
+  for (set<Edge*>::iterator de = deplog_edges.begin(); de != deplog_edges.end();
+       ++de) {
+    if (!state_->PathExistsBetween(*de, edge)) {
+      missing_deps.push_back(*de);
+    }
+  }
+
+  for (vector<Node*>::iterator ni = deplog_nodes_without_generator.begin();
+       ni != deplog_nodes_without_generator.end(); ++ni) {
+    printf("No generator: %s uses %s, but I don't know who generates that.\n",
+         node->path().c_str(), (*ni)->path().c_str());
+  }
+  if (!missing_deps.empty()) {
+    set<string> missing_deps_rule_names;
+    for (vector<Edge*>::iterator ne = missing_deps.begin();
+         ne != missing_deps.end(); ++ne) {
+      for (int i = 0; i < dep_nodes_count; ++i) {
+        if (dep_nodes[i]->in_edge() == *ne) {
+          generated_nodes_.insert(dep_nodes[i]);
+          printf("Missing dep: %s uses %s (generated by %s)\n",
+              node->path().c_str(), dep_nodes[i]->path().c_str(),
+              (*ne)->rule().name().c_str());
```

Comment:
indent L630~L631

Reply:
obsolete after code move to src/missing_deps.cc